### DISCUSSION TITLE
How to prevent executable_ from being added to Python files on Windows?

### DISCUSSION THREAD

**Discussion by eugenesvk** [twpayne/chezmoi#3777] [Q&A]
If I have a Mac/Win python config file that doesn't need to be executable, so when added on a Mac it doesn't have `executable_` flag

But then if I edit it on a Windows and re-add it, the flag is also added

How do I avoid that?

Thanks!

**Comment by twpayne** [comment]
> But then if I edit it on a Windows and re-add it, the flag is also added

This is because Windows always adds the executable flag.



> How do I avoid that?

Use `chezmoi edit` instead of `chezmoi re-add`.

**Reply by eugenesvk** [reply]
Thanks, but unfortunately many configs are edited by other apps, so by the time I get to the syncing part `edit` would have an outdated view

**Reply by twpayne** [reply]
In that case you might be able to use a [hook](https://www.chezmoi.io/reference/configuration-file/hooks/), say `hooks.re-add.post` to remove the `executable_` attribute in the source directory.

**Reply by eugenesvk** [reply]
Will I have enough data for the hook to work? I'd like to remove it only if it was set, but then I don't know what the result of chezmoi operation is, I only get pre-operations `CHEZMOI_ARGS`, right?

Is it possible to add a flag to `re-add` to suppress changing permission bits? Then I could set that command as my default since I almost never need to change permissions after the initial setup?

**Reply by twpayne** [reply]
> Will I have enough data for the hook to work?

Yes, you can make arbitrary modifications to the source directory.



> Is it possible to add a flag to `re-add` to suppress changing permission bits?

No, this is not easy to add.

**Reply by eugenesvk** [reply]
> Yes, you can make arbitrary modifications to the source directory.

The issue is **whether** to make any, e.g., if there is a legitimate `executable_` file, I don't want to remove that flag even if this path was passed to chezmoi as an argument, but **only** if chezmoi itself has renamed an existing file and appended `executable_` bit to it (and only on Windows), hence the question about what **data** from chezmoi is available to the hook

**Reply by twpayne** [reply]
This doesn't need to be complicated:

```bash
#!/bin/bash

# remove the `executable_` attribute which was added when something else modified the file
if [ -f ${CHEZMOI_SOURCE_DIR}/executable_dot_file ] ; then
    mv ${CHEZMOI_SOURCE_DIR}/executable_dot_file ${CHEZMOI_SOURCE_DIR}/dot_file
fi
```

If you need to modify more than one file then you can use a `for` loop or `find`.

**Reply by eugenesvk** [reply]
But you don't know whether this attribute was added! So this will just blindly remove all the executable bits

**Reply by twpayne** [reply]
No, I assume that you know which files are affected by the permissions corruption on your system. Only apply the fix to those files.

**Reply by eugenesvk** [reply]
This is mostly a silent bug, so I don't really have a list as that can be any random file with the right extension. I see now that there are plenty of errouneous `executable_` bits set for various py/js scripts, but then many of these weren't x-platform, so I never noticed, or they were, but were created on Windows first, so the exe bit on non-Win was just set and left unchanged, so there was no swapping to get my attention

Maybe I'll just do the same and mark these "swapping" files as executable and live with the extra flag, easier than having to keep and update a list of files for the hook...
